kuri_camera/src/h264_decoder.cpp

```diff
@@ -0,0 +1,348 @@
+    cv_linesizes[0] = image.step1();
+
+    // Convert the color format and write directly to the opencv matrix
+    SwsContext* conversion = sws_getContext(width, height, /*(AVPixelFormat) frame->format*/AV_PIX_FMT_YUV420P, width, height, AV_PIX_FMT_BGR24 /* PIX_FMT_BGR24 */, SWS_FAST_BILINEAR, NULL, NULL, NULL);
+    sws_scale(conversion, frame->data, frame->linesize, 0, height, &image.data, cv_linesizes); // copies data
+    sws_freeContext(conversion);
+}
+
+void H264Decoder::convertToMat() {
+  while (is_alive) {
+    std::unique_lock<std::mutex> picture_lock(picture_mutex);
+    got_picture_cvar.wait(picture_lock);
+    std::unique_lock<std::mutex> cv_image_lock(cv_image_mutex);
+    if (got_picture) {
+      avframeToMat(picture, cv_image);
+      cv_image_recv_timestamp = picture_recv_timestamp;
+      cv_image_lock.unlock();
+      decoded_new_frame.notify_all();
+    }
+    picture_lock.unlock();
+  }
+}
+
+int H264Decoder::getMostRecentFrame(cv::Mat &image, uint64_t &recv_timestamp) {
+  std::unique_lock<std::mutex> cv_image_lock(cv_image_mutex);
+  image = cv_image.clone();
+  recv_timestamp = cv_image_recv_timestamp;
+  cv_image_lock.unlock();
+}
+
+bool H264Decoder::load(std::string hostname, std::string port) {
+
+  codec = avcodec_find_decoder(AV_CODEC_ID_H264);
+  if(!codec) {
+    printf("Error: cannot find the h264 codec: %s:%s\n", hostname.c_str(), port.c_str());
+    return false;
+  }
+
+  codec_context = avcodec_alloc_context3(codec);
+
+  if(codec->capabilities & CODEC_CAP_TRUNCATED) {
+    codec_context->flags |= CODEC_FLAG_TRUNCATED;
+  }
+
+  if(avcodec_open2(codec_context, codec, NULL) < 0) {
+    printf("Error: could not open codec.\n");
+    return false;
+  }
+
+  // Get the TCP endpoint
+  boost::asio::ip::tcp::resolver resolver(io_service);
+  boost::asio::ip::tcp::resolver::query query(hostname, port);
+  iter = resolver.resolve(query);
+
+  std::unique_lock<std::mutex> picture_lock(picture_mutex);
+  // NOTE: depending on your version of libav, you may need the first or second
+  // function below to allocate a frame.
+  picture = av_frame_alloc(); // avcodec_alloc_frame(); // 
+  picture_lock.unlock();
+  parser = av_parser_init(AV_CODEC_ID_H264);
+  packet = new AVPacket;
+  av_init_packet(packet);
+
+  if(!parser) {
+    printf("Erorr: cannot create H264 parser.\n");
+    return false;
+  }
+
+  return true;
+}
+
+void H264Decoder::readFrame() {
+  while (is_alive) {
+    std::unique_lock<std::mutex> got_frame_lock(got_frame_mutex);
+    got_frame.wait(got_frame_lock);
+
+    std::unique_lock<std::mutex> buffer_lock(buffer_mutex);
+    if (buffer_frame_begin_indices.size() > 1) {
+
+      // Only process the most recent frame (to prevent buffer overflow)
+      size_t index0 = buffer_frame_begin_indices[buffer_frame_begin_indices.size()-2];
+      size_t index1 = buffer_frame_begin_indices[buffer_frame_begin_indices.size()-1];
+
+      int decodedAmount = decodeFrame(&buffer[index0], index1-index0);
```

Comment:
I know it's from borrowed code, but consistency :smiley:

Reply:
Haha that problem is actually not due to the borrowed code. I am so used to camel case that it is second-nature whenever I write code. I'm actively working to undo that, but instances like this crop up :)